Pass COMPUTERNAME env var to elasticsearch.bat

williamrandolph opened the pull request:
When we run bin/elasticsearch with bash, we get a $HOSTNAME builtin that
contains the hostname of the machine the script is running on. When
there's no provided nodename, Elasticsearch uses the HOSTNAME to create
a nodename. On Windows, Powershell provides a $COMPUTERNAME variable for
the same purpose. CMD.EXE provides the same thing, except it's called
%COMPUTERNAME%. bin/elasticsearch.bat sets $HOSTNAME to the value of
$COMPUTERNAME. However, when testclusters invokes bin/elasticsearch.bat,
the COMPUTERNAME variable doesn't get passed in, leaving HOSTNAME null
and breaking an integration test on Windows.

This commit sets COMPUTERNAME in the environment so that our tests get
the value that Elasticsearch would have when bin/elasticsearch.bat is
invoked from the shell.

Fixes #44656

williamrandolph committed:
Pass COMPUTERNAME env var to elasticsearch.bat

When we run bin/elasticsearch with bash, we get a $HOSTNAME builtin that
contains the hostname of the machine the script is running on. When
there's no provided nodename, Elasticsearch uses the HOSTNAME to create
a nodename. On Windows, Powershell provides a $COMPUTERNAME variable for
the same purpose. CMD.EXE provides the same thing, except it's called
%COMPUTERNAME%. bin/elasticsearch.bat sets $HOSTNAME to the value of
$COMPUTERNAME. However, when testclusters invokes bin/elasticsearch.bat,
the COMPUTERNAME variable doesn't get passed in, leaving HOSTNAME null
and breaking an integration test on Windows.

This commit sets COMPUTERNAME in the environment so that our tests get
the value that Elasticsearch would have when bin/elasticsearch.bat is
invoked from the shell.

williamrandolph committed:
Add null check to protect in non-Windows case

What good is it a developer to gain the whole Windows if they forfeit
their Unix? The value that fixes things on Windows is null on
Linux/Darwin, so let's null-check it.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -642,6 +642,10 @@ private void runElaticsearchBinScript(String tool, String... args) {
         // Windows requires this as it defaults to `c:\windows` despite ES_TMPDIR
         defaultEnv.put("TMP", tmpDir.toString());
+        // Windows needs the COMPUTERNAME variable in order to set HOSTNAME
+        if (System.getenv("COMPUTERNAME") != null) {
```
Where is HOSTNAME passed through for linux? Do we actually need to rely on the real system computername or could we pass in a dummy one for tests?

williamrandolph replied:
HOSTNAME is a builtin for Bash, so we don't need to pass it through. I believe it's always available unless you explicitly unset it:

https://www.gnu.org/software/bash/manual/html_node/Bash-Variables.html

I think a dummy value will work and avoid the need for a null check.

williamrandolph replied:
In fact, the integration test that was failing here just checks that the node name is not an empty string. I'm going to pass in HOSTNAME and COMPUTERNAME explicitly and test for specific values.

williamrandolph committed:
Override system hostnames for testclusters

Rather than relying on variable system behavior, let's just override
HOSTNAME and COMPUTERNAME and test for correct values in the integration
test that was originally failing.

rjernst reviewed (approved):
LGTM

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -107,6 +107,8 @@
         "is a pre-release version of Elasticsearch",
         "max virtual memory areas vm.max_map_count"
     );
+    private static final String LINUX_DARWIN_HOSTNAME_OVERRIDE = "LinuxDarwinHostname";
```
Is there any value  in providing different values per platform here ? Could we provide one and avoid platform dependent checks in the test ?

williamrandolph replied:
Mainly I wanted to convince myself that I understood how the batch scripts set the hostname variable. For Windows, you might think that setting a `$HOSTNAME` environment variable would do the trick, since that's what the bootstrap code ultimately reads, but that would be wrong, since `elasticsearch.bat` will overwrite `$HOSTNAME` with the value in `$COMPUTERNAME` before starting the application.

Since the different platforms determine the host name differently, I think there's value in demonstrating that in an integration test. On the other hand, I see the value in platform independence for these tests (as opposed to the specifically platform-dependent OS tests). Let me know what you think. I will be happy to change this if you'd like me to (and @rjernst, our other reviewer, agrees).

atorok replied:
That's a good point. We also test that HOSTNAME doesn't influence Windows this way. 
I would frop "LIUX_DARWIN" and "WINDOWS" from the constant names since we always set them.

williamrandolph committed:
Rename constants for clarity

Since we are setting HOSTNAME and COMPUTERNAME regardless of whether the
tests are running on Windows or Linux, we shouldn't imply that constants
are only used in one case or the other.

williamrandolph commented:
@elasticmachine Please run elasticsearch-ci/2

alpar-t reviewed (approved):
LGTM! Thanks!

jakelandis commented:
moved release label to 7.3.2 since it just barely missed the 7.3.1 cutoff.
